src/Analyzer/Passes/IfTransformStringsToEnumPass.cpp

```diff
@@ -0,0 +1,306 @@
+    {
+        auto cast_function = FunctionFactory::instance().get("_CAST", context);
+        QueryTreeNodes arguments{node, enum_literal_node};
+
+        auto function_node = std::make_shared<FunctionNode>("_CAST");
+        function_node->resolveAsFunction(std::move(cast_function), result_type);
+        function_node->getArguments().getNodes() = std::move(arguments);
+
+        node = std::move(function_node);
+    };
+
+    create_cast_function(first);
+    create_cast_function(second);
+}
+
+void changeTransformArguments(
+    QueryTreeNodePtr & array_to,
+    QueryTreeNodePtr & default_value,
+    const std::set<std::string> & string_values,
+    const ContextPtr & context)
+{
+    auto [result_type, enum_string] = getEnumTypeAndString(string_values);
+
+    {
+        auto enum_literal = std::make_shared<ConstantValue>(fmt::format("Array({})", enum_string), std::make_shared<DataTypeString>());
+        auto enum_literal_node = std::make_shared<ConstantNode>(std::move(enum_literal));
+
+        auto cast_function = FunctionFactory::instance().get("_CAST", context);
+        QueryTreeNodes arguments{array_to, enum_literal_node};
+
+        auto function_node = std::make_shared<FunctionNode>("_CAST");
+        function_node->resolveAsFunction(std::move(cast_function), std::make_shared<DataTypeArray>(result_type));
+        function_node->getArguments().getNodes() = std::move(arguments);
+
+        array_to = std::move(function_node);
+    }
+
+    {
+        auto enum_literal = std::make_shared<ConstantValue>(enum_string, std::make_shared<DataTypeString>());
+        auto enum_literal_node = std::make_shared<ConstantNode>(std::move(enum_literal));
+
+        auto cast_function = FunctionFactory::instance().get("_CAST", context);
+        QueryTreeNodes arguments{default_value, enum_literal_node};
+
+        auto function_node = std::make_shared<FunctionNode>("_CAST");
+        function_node->resolveAsFunction(std::move(cast_function), result_type);
+        function_node->getArguments().getNodes() = std::move(arguments);
+
+        default_value = std::move(function_node);
+    }
+}
+
+class ConvertStringsToEnumVisitor : public InDepthQueryTreeVisitor<ConvertStringsToEnumVisitor>
+{
+public:
+    explicit ConvertStringsToEnumVisitor(std::unordered_set<std::string_view> used_functions_, ContextPtr context_)
+        : used_functions(std::move(used_functions_)), context(std::move(context_))
+    {
+    }
+
+    static bool needChildVisit(VisitQueryTreeNodeType & parent, VisitQueryTreeNodeType & /* child */)
+    {
+        return parent->getNodeType() != QueryTreeNodeType::FUNCTION;
+    }
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        auto * function_node = node->as<FunctionNode>();
+
+        if (!function_node)
+            return;
+
+        /// we cannot change the type of its result because it's used
+        /// as argument in another function
+        if (used_functions.contains(function_node->getAlias()))
```

Comment:
There should be no aliases at this stage. Probably it is okay to use pointers here.

Reply:
Actually I double checked and we can just clone this node, if we modify its content, that way we do not need to think if it is used inside any other functions at all.
Althought even in this case such optimization can be tricky, and can easily break something.
For example `SELECT if(arg1, arg2, arg3)`. If such query will be modified, we must change query node header. But it is not the worst scenario. Consider `SELECT t1.value FROM (SELECT if(arg1, arg2, arg3) AS value) AS t1`. 

As solution I propose that we clone `if`, `transform` function if we modify it, and also wrap it into `toString`.
```SELECT concat(toString(if(number > 5, _CAST('censor.net', 'Enum8(\'censor.net\' = 1, \'google\' = 2)'), _CAST('google', 'Enum8(\'censor.net\' = 1, \'google\' = 2)'))), '1') FROM numbers(10)```. Althought it is hard to predict if this optimization will make any sense after that, need to make a couple of tests.

Reply:
why is `concat` necessary in this case? AFAIU, the point is to keep the type as string right, shouldn't `toString` be enough then?

Reply:
Yes, concat is just example. I think `toString` will be good. For example for transform it still makes a lot of sense, because it builds hash table internally, and if we pass enum it will build hash table of integers, should work much faster.

Reply:
Okay, will try to do that.

Also, what would be the point of cloning?
Why can't I just take the initial `FunctionNode` (e.g. 'if' function) and pass it as an argument to `toString` `FunctionNode`?

Reply:
You can do this, but this can be tricky.
In such example: ```SELECT if(number > 5, 'a', 'b') AS value, value```.
After query analysis pass is should be resolved like: ```SELECT if(number > 5, 'a', 'b'), if(number > 5, 'a', 'b')```. But you need to keep in mind that this first `if` ans second `if` function point to the same function node, or not. It is implementation defined, so we should not rely on in. In current implementation they will point to the same function node.
If you modify first ```if```, the query tree should look like this: ```SELECT if(number > 5, _CAST('a', 'Enum8(\'a\' = 1, \'b\' = 2)', _CAST('b', 'Enum8(\'a\' = 1, \'b\' = 2)'), if(number > 5, _CAST('a', 'Enum8(\'a\' = 1, \'b\' = 2)', _CAST('b', 'Enum8(\'a\' = 1, \'b\' = 2)')```.
But if you replace `if` into `toString` with `if` function as argument, you will wrap only first `if`, and the second `if` will be invalid.
We should not just replace node, but resolve `if` function as `toString` function, where initial `if` is `toString` function argument. That way all function nodes will be replaced automatically.